app/migrate.go

```diff
@@ -0,0 +1,202 @@
+			if err = MigrateOldVolume(kubeClient, lhClient, lhNamespace, v.Name); err != nil {
+				return err
+			}
+		}
+	} else {
+		if c.NArg() == 0 {
+			return errors.New("volume name or the flag '--all' is required")
+		}
+		if err = MigrateOldVolume(kubeClient, lhClient, lhNamespace, c.Args()[0]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func MigrateOldVolume(kubeClient *kubeclientset.Clientset, lhClient *lhclientset.Clientset, lhNamespace, volumeName string) (err error) {
+	defer func() {
+		if err != nil {
+			err = errors.Wrapf(err, "Failed to migrate the old volume %v", volumeName)
+		}
+	}()
+
+	v, err := lhClient.LonghornV1beta1().Volumes(lhNamespace).Get(volumeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	ks := v.Status.KubernetesStatus
+
+	if v.Status.State != types.VolumeStateDetached {
+		logrus.Infof("Invalid state %v for migrating volume %v", v.Status.State, volumeName)
+		return nil
+	}
+	if ks.PVName == "" {
+		logrus.Infof("There is no need to migrate volume %v: no related PV", volumeName)
+		return nil
+	}
+	if len(ks.WorkloadsStatus) != 0 && ks.LastPodRefAt == "" {
+		logrus.Infof("There are still running workloads using the volume %v", volumeName)
+		return nil
+	}
+
+	oldPV, err := kubeClient.CoreV1().PersistentVolumes().Get(ks.PVName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if oldPV.Spec.CSI == nil || oldPV.Spec.CSI.Driver != types.DepracatedDriverName {
+		logrus.Infof(" There is no need to migrate volume %v: the PV is not created by the old Longhorn", volumeName)
+		return nil
+	}
+
+	// Recreate PVC and PV without using CSI/StorageClass can make things easier:
+	// 1) The goal of the volume migration is deprecating the old CSI components in Longhorn system.
+	// As long as the data is retained and the volume can be controlled by the new CSI components,
+	// the inconsistency of some PVC/PV fields do not matter.
+	// 2) The StorageClass is used for volume initialization and setting some immutable fields in PV/PVC.
+	// But currently the volume already exists and the parameters may be different from those immutable field.
+	// Hence we don't need to make sure that those fields in the new PV/PVC are the same as the old ones.
+	// pv.Spec.CSI.VolumeAttributes
+	// 3) If we choose to use the StorageClass, we need to make sure:
+	//     1) the StorageClass used by the old PV and PVC still exists;
+	//     2) its parameters are the same as oldPV.Spec.CSI.VolumeAttributes.
+	staticStorageClass, err := lhClient.LonghornV1beta1().Settings(lhNamespace).Get(
+		string(types.SettingNameDefaultLonghornStaticStorageClass), metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if staticStorageClass.Value == "" {
+		return fmt.Errorf("empty static StorageClass for new PV/PVC creation")
+	}
+
+	if oldPV.Spec.PersistentVolumeReclaimPolicy != v1.PersistentVolumeReclaimRetain {
+		oldPV.Spec.PersistentVolumeReclaimPolicy = v1.PersistentVolumeReclaimRetain
```

Comment:
PV's `retain` only refer to not delete the PV automatically when PVC deleted to my understanding, though I noticed we won't delete the volume from Longhorn as well. Can you point me to the code that skip the deletion? Just want to make sure it's safe to delete the PV.

Reply:
[This doc](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retain) describes the behavior:
```
Retain:
Delete the PersistentVolume. The associated storage asset in external infrastructure (such as an AWS EBS, GCE PD, Azure Disk, or Cinder volume) still exists after the PV is deleted.

Delete:
For volume plugins that support the Delete reclaim policy, deletion removes both the PersistentVolume object from Kubernetes, as well as the associated storage asset in the external infrastructure, such as an AWS EBS, GCE PD, Azure Disk, or Cinder volume.
```

Reply:
This looks good.